<a href="https://colab.research.google.com/github/pponali/masai/blob/main/Leakage_Imbalance.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [28]:
!pip3 install pandas
!pip3 install scikit-learn

# path of the file /content/vehicles.csv

# Problem Statement

##### Task 1 — Explore and Identify Issues Load the dataset and use df.info(), df.describe(), and df.shape to report at least three data quality problems you observe (e.g., wrong dtypes, nulls, impossible values).

##### Task 2 — Clean the Data Fix the issues identified: drop null target rows, impute missing input features, strip and lowercase the brand column, extract numeric values from the mileage column, and remove duplicates.

##### Task 3 — Compute Baseline MAE Build a baseline model that predicts the mean selling_price for every record. Calculate and print the MAE of this baseline on the full cleaned dataset.



In [39]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

# Task 1 — Explore and Identify Issues Load the dataset and use df.info(), df.describe(), and df.shape to report at least three data quality problems you observe (e.g., wrong dtypes, nulls, impossible values).

In [15]:
df = pd.read_csv("/content/cardekho_dataset 2.csv")
df

,Unnamed: 0,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15406,19537,Hyundai i10,Hyundai,i10,9,10723,Dealer,Petrol,Manual,19.81,1086,68.05,5,250000
15407,19540,Maruti Ertiga,Maruti,Ertiga,2,18000,Dealer,Petrol,Manual,17.50,1373,91.10,7,925000
15408,19541,Skoda Rapid,Skoda,Rapid,6,67000,Dealer,Diesel,Manual,21.14,1498,103.52,5,425000
15409,19542,Mahindra XUV500,Mahindra,XUV500,5,3800000,Dealer,Diesel,Manual,16.00,2179,140.00,7,1225000


In [16]:
df.drop(columns=['Unnamed: 0'], inplace=True)
df

,car_name,brand,model,vehicle_age,km_driven,seller_type,fuel_type,transmission_type,mileage,engine,max_power,seats,selling_price
0,Maruti Alto,Maruti,Alto,9,120000,Individual,Petrol,Manual,19.70,796,46.30,5,120000
1,Hyundai Grand,Hyundai,Grand,5,20000,Individual,Petrol,Manual,18.90,1197,82.00,5,550000
2,Hyundai i20,Hyundai,i20,11,60000,Individual,Petrol,Manual,17.00,1197,80.00,5,215000
3,Maruti Alto,Maruti,Alto,9,37000,Individual,Petrol,Manual,20.92,998,67.10,5,226000
4,Ford Ecosport,Ford,Ecosport,6,30000,Dealer,Diesel,Manual,22.77,1498,98.59,5,570000
...,...,...,...,...,...,...,...,...,...,...,...,...,...
15406,Hyundai i10,Hyundai,i10,9,10723,Dealer,Petrol,Manual,19.81,1086,68.05,5,250000
15407,Maruti Ertiga,Maruti,Ertiga,2,18000,Dealer,Petrol,Manual,17.50,1373,91.10,7,925000
15408,Skoda Rapid,Skoda,Rapid,6,67000,Dealer,Diesel,Manual,21.14,1498,103.52,5,425000
15409,Mahindra XUV500,Mahindra,XUV500,5,3800000,Dealer,Diesel,Manual,16.00,2179,140.00,7,1225000


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15411 entries, 0 to 15410
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   car_name           15411 non-null  object 
 1   brand              15411 non-null  object 
 2   model              15411 non-null  object 
 3   vehicle_age        15411 non-null  int64  
 4   km_driven          15411 non-null  int64  
 5   seller_type        15411 non-null  object 
 6   fuel_type          15411 non-null  object 
 7   transmission_type  15411 non-null  object 
 8   mileage            15411 non-null  float64
 9   engine             15411 non-null  int64  
 10  max_power          15411 non-null  float64
 11  seats              15411 non-null  int64  
 12  selling_price      15411 non-null  int64  
dtypes: float64(2), int64(5), object(6)
memory usage: 1.5+ MB


In [18]:
df.describe()

,vehicle_age,km_driven,mileage,engine,max_power,seats,selling_price
count,15411.000000,1.541100e+04,15411.000000,15411.000000,15411.000000,15411.000000,1.541100e+04
mean,6.036338,5.561648e+04,19.701151,1486.057751,100.588254,5.325482,7.749711e+05
std,3.013291,5.161855e+04,4.171265,521.106696,42.972979,0.807628,8.941284e+05
min,0.000000,1.000000e+02,4.000000,793.000000,38.400000,0.000000,4.000000e+04
25%,4.000000,3.000000e+04,17.000000,1197.000000,74.000000,5.000000,3.850000e+05
50%,6.000000,5.000000e+04,19.670000,1248.000000,88.500000,5.000000,5.560000e+05
75%,8.000000,7.000000e+04,22.700000,1582.000000,117.300000,5.000000,8.250000e+05
max,29.000000,3.800000e+06,33.540000,6592.000000,626.000000,9.000000,3.950000e+07


In [20]:
df.shape

(15411, 13)

 #  Report at least three data quality problems you observe (e.g., wrong dtypes, nulls, impossible values).


 | Column No. | Column Name        | Non-Null Count | Data Type |
|------------|--------------------|----------------|-----------|
| 0          | car_name           | 15411          | object    |
| 1          | brand              | 15411          | object    |
| 2          | model              | 15411          | object    |
| 3          | vehicle_age        | 15411          | int64     |
| 4          | km_driven          | 15411          | int64     |
| 5          | seller_type        | 15411          | object    |
| 6          | fuel_type          | 15411          | object    |
| 7          | transmission_type  | 15411          | object    |
| 8          | mileage            | 15411          | float64   |
| 9          | engine             | 15411          | int64     |
| 10         | max_power          | 15411          | float64   |
| 11         | seats              | 15411          | int64     |
| 12         | selling_price      | 15411          | int64     |


# Based on this data there is no wrong datatype in the given data set.

# No Null values, all the columns has 15411 rows.

# Impossible values.

 ### vehicle age can not be 0 for used cars.
 ### number of seats can not be zero for any car.



# Task 2 — Clean the Data Fix the issues identified: drop null target rows, impute missing input features, strip and lowercase the brand column, extract numeric values from the mileage column, and remove duplicates.


## With this data set no need of drop nulls and impute the missing input features.




# Strip and Lowercase the brand column,

In [24]:
df['brand'] = df['brand'].str.strip()
df['brand']

,brand
0,maruti
1,hyundai
2,hyundai
3,maruti
4,ford
...,...
15406,hyundai
15407,maruti
15408,skoda
15409,mahindra


In [23]:
df['brand'] = df['brand'].str.lower()
df['brand']

,brand
0,maruti
1,hyundai
2,hyundai
3,maruti
4,ford
...,...
15406,hyundai
15407,maruti
15408,skoda
15409,mahindra


#  Extract numeric values from the mileage column

### Ans: data is already clean. no need to extract numerical values from the mileage. Since data type is int64.

 # Remove duplicates.

In [25]:
df.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
15406,False
15407,False
15408,False
15409,False


### Ans:  No duplicates in the given data set.

# Task 3 — Compute Baseline MAE Build a baseline model that predicts the mean selling_price for every record. Calculate and print the MAE of this baseline on the full cleaned dataset.

In [30]:
df.columns

Index(['car_name', 'brand', 'model', 'vehicle_age', 'km_driven', 'seller_type',
       'fuel_type', 'transmission_type', 'mileage', 'engine', 'max_power',
       'seats', 'selling_price'],
      dtype='object')

In [59]:

x = df [['car_name', 'brand', 'model', 'vehicle_age', 'km_driven',
       'seller_type', 'fuel_type', 'transmission_type', 'mileage', 'engine',
       'max_power', 'seats']]

y = df['selling_price']

x_temp, x_test, y_temp, y_test = train_test_split(x,y, train_size=.80, random_state=42)

x_train, x_val, y_train, y_val = train_test_split(x_temp, y_temp, test_size=0.18, random_state=42)


In [62]:
base_line_prediction = np.mean(y_train)
base_line_predictions = np.full(len(y_val), base_line_prediction)

In [64]:
from sklearn.metrics import mean_absolute_error

baseline_mae = mean_absolute_error(y_val, base_line_predictions)
print(f"Baseline MAE : {baseline_mae}")

Baseline MAE : 449486.6037782935
